### PR TITLE
feat: add negative sampler and  MultiTaskDatasetManager

### PR DESCRIPTION
**Pull request by Zoe2021Wang**
This pull request introduces two key features:
1. NegativeSampler
- Implements a negative sampling strategy based on length bins.
- Designed to support both single-task and multi-task peptide datasets.
2.MultiTaskDatasetManager
- Extends the existing dataset manager to handle multi-label scenarios.

### Changes
- Added `NegSampler` and `MultiTaskNegSampler` classes in `pep_utils/neg_sample.py`.
- Created `MultiTaskDatasetManager` in `single_pred/multitask_dataset.py`.
- Refactored the dataset interface to allow integration with negative sampling pipelines.
-

### CODE REVIEW

File: src/pepbenchmark/metadata.py
Comment (Kqiii) [inline src/pepbenchmark/metadata.py:1]: 为什么binary_classification的num_class是1？我大概能懂这个意思但是”二分类的类别数是1“这听起来有一点奇怪。"nature": "natural"这个 kv pair的语义有重复, 建议修改：如果仅有两类，那么改成"natural": True/False. 如果有多类，那么改成"xxx_type": "nature"/"non_nature"/something else

File: src/pepbenchmark/pep_utils/neg_sample.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    def __init__(self, dataset_name: str):
+        if dataset_name in EXCLUSIVE_MAP:
+            self.select = "exclusive"
+            self.mapping = EXCLUSIVE_MAP[dataset_name]
+        elif dataset_name in INCLUDE_MAP:
+            self.select = "inclusive"
+            self.mapping = INCLUDE_MAP[dataset_name]
+        else:
+            raise ValueError(
+                "Negative sampling is not supported for the given dataset name."
```
Comment (Kqiii) [inline src/pepbenchmark/pep_utils/neg_sample.py:46]: 补全提示: f"Negative sampling is not supported for the {dataset_name} dataset."

File: src/pepbenchmark/pep_utils/neg_sample.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+            )
+
+    def __call__(self, pos_seq, ratio, seed):
+        self.seed = seed
+        return self._sample_from_bioactive(pos_seq, ratio, self.select, self.mapping)
```
Comment (Kqiii) [inline src/pepbenchmark/pep_utils/neg_sample.py:51]: __call__是预计后续会扩展功能吗？如果不会，请把_sample_from_bioactive的逻辑挪到__call__里，能少包一层

File: src/pepbenchmark/pep_utils/neg_sample.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+                shortage -= borrow_count
+                step += 1
+
+            neg_samples.append(sampled)
+
+        neg_df = pd.concat(neg_samples).drop_duplicates(subset="sequence")
+        return neg_df["sequence"].tolist()
+
+
+class MultiTaskNegSampler(NegSampler):
```
Comment (Kqiii) [inline src/pepbenchmark/pep_utils/neg_sample.py:134]: 继承了NegSampler但是没有调用super(), please make sure this is an expected behaviour

File: src/pepbenchmark/pep_utils/neg_sample.py
Comment (Kqiii) [inline src/pepbenchmark/pep_utils/neg_sample.py:1]: 这个文件的typing hint不完整

File: src/pepbenchmark/single_pred/multitask_dataset.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+LABEL_TYPE = ["Antibacterial", "Antiviral"]
+
+BASE_URL = "https://raw.githubusercontent.com/ZGCI-AI4S-Pep/peptide_data/main/"
+
+
+class MultiTaskDatasetManager(DatasetManager):
+    def __init__(
+        self,
+        dataset_name: str = "multitask_peptidepedia",
+        labels: List[str] = None,  # activity names, et.al. []
```
Comment (Kqiii) [inline src/pepbenchmark/single_pred/multitask_dataset.py:40]: Then labels are optional: Optional[List[str]] ?

File: src/pepbenchmark/single_pred/singeltask_dataset.py [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+            with requests.get(url, stream=True, timeout=100) as r:
+                r.raise_for_status()
+                with open(feature_path, "wb") as f:
+                    for chunk in r.iter_content(chunk_size=8192):
+                        f.write(chunk)
+        except requests.RequestException as e:
+            logger.error(f"Failed to download {feature_name}: {e}")
+            raise
+
+    def negative_sampling(self, ratio: int, **kwargs):
```
Comment (Kqiii) [inline src/pepbenchmark/single_pred/singeltask_dataset.py:268]: ratio不能是int

File: src/pepbenchmark/single_pred/multitask_dataset.py
Comment (Kqiii) [inline src/pepbenchmark/single_pred/multitask_dataset.py:1]: 为什么multitask_dataset.py这个文件会在single_pred这个文件夹里？single_pred是说什么东西是single的？

File: notebook/basic.ipynb
[the remark this answers is not in the document]
Comment (zjiahui599) [reply notebook/basic.ipynb:1]: 这个不作为包的内容，是用来写专门写使用文档的。

File: src/pepbenchmark/single_pred/multitask_dataset.py
Comment (Zoe2021Wang) [reply src/pepbenchmark/single_pred/multitask_dataset.py:1]: 这里single指的是肽不是任务，放在这个文件夹的数据集方法都是针对单个肽分子性质预测的，与之对应的有肽-蛋白结合、肽-MHC分子结合的数据集，但这个命名确实有点歧义，后面改成single_peptide之类的吧

File: src/pepbenchmark/single_pred/multitask_dataset.py [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+        test_features.update(_collect_features(self.user_feature_dict, test_idx))
+
+        if format == "dict":
+            return train_features, valid_features, test_features
+        elif format == "pytorch_dataset":
+
+            class PyDataset(Dataset):
+                def __init__(self, features: Dict[str, Any]):
+                    self.features = features
+
```
Comment (Minys233) [inline src/pepbenchmark/single_pred/multitask_dataset.py:277]: For `Dataset` class, `__len__` method should be the number of the data points included, not the feature dimension of the tensor.

File: notebook/basic.ipynb
[the remark this answers is not in the document]
Comment (Kqiii) [reply notebook/basic.ipynb:1]: > 这个不作为包的内容，是用来写专门写使用文档的。

好的，那可以先留着。后续删掉吧

File: src/pepbenchmark/single_pred/multitask_dataset.py
Comment (Zoe2021Wang) [reply src/pepbenchmark/single_pred/multitask_dataset.py:1]: 好的，这次就可以改

File: src/pepbenchmark/pep_utils/neg_sample.py (hunk 2 above)
Comment (Zoe2021Wang) [reply src/pepbenchmark/pep_utils/neg_sample.py]: 已将_sample_from_bioactive的逻辑移到__call__里了

File: src/pepbenchmark/pep_utils/neg_sample.py (hunk 3 above)
Comment (Zoe2021Wang) [reply src/pepbenchmark/pep_utils/neg_sample.py]: 多任务数据集相关功能后续可能与单任务合并，这次先没有改

File: src/pepbenchmark/single_pred/singeltask_dataset.py [hunk 7]
```diff
@@ -0,0 +1,463 @@
...
+            need_download = True
+
+        if need_download:
+            self._download_official_feature(feature_name)
+
+        if file_extention == "csv":
+            feature = pd.read_csv(feature_path)["feature"]
+        elif file_extention == "npz":
+            feature = np.load(feature_path)["data"]
+        elif file_extention == "pt":
```
[the remark this answers is not in the document]
Comment (Zoe2021Wang) [reply src/pepbenchmark/single_pred/singeltask_dataset.py]: 这里的图数据是torch.tensor的形式

File: src/pepbenchmark/single_pred/singeltask_dataset.py [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+        if format == "dict":
+            return train_features, valid_features, test_features
+        elif format == "pytorch_dataset":
+
+            class PyDataset(Dataset):
+                def __init__(self, features: Dict[str, Any]):
+                    self.features = features
+
+                def __len__(self):
+                    return len(next(iter(self.features.values())))
```
[the remark this answers is not in the document]
Comment (Zoe2021Wang) [reply src/pepbenchmark/single_pred/singeltask_dataset.py]: 这个计算的应该是第一个feature的长度，也就是data point的数量

File: src/pepbenchmark/single_pred/singeltask_dataset.py (hunk 5 above)
Comment (Kqiii) [reply src/pepbenchmark/single_pred/singeltask_dataset.py]: > 这个方法改为def set_negative_samples(self, neg_seqs: List[str])了，typing hint 在class NegSampleFromBioactive的__call__中修改了

确认一下：不是简单地改了type哈？ratio真的是当做“比例”而不是“数量”用的对吧？

File: src/pepbenchmark/single_pred/singeltask_dataset.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Kqiii) [reply src/pepbenchmark/single_pred/singeltask_dataset.py]: 所以迭代的是feature哈？就是说每次next()返回的都是一个(num_samples, feat_dim)这样的东西？理解了。请把这个信息作为comment加在这条语句出现的地方，因为目前看起来这个写法比较容易引起误解。

File: src/pepbenchmark/single_pred/singeltask_dataset.py (hunk 5 above)
Comment (Zoe2021Wang) [reply src/pepbenchmark/single_pred/singeltask_dataset.py]: 是的，这里ratio传入的是比例

File: src/pepbenchmark/pep_utils/neg_sample.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+    "QS_APML": [],
+    "Antibacterial": ["Antibacterial"],
+    "Antiviral": ["Antiviral"],
+}  # TODO: To be complete
+
+INCLUDE_MAP = {
+    "Nonfouling": ["Hemolytic", "Antimicrobial"],
+}  # TODO: To be complete
+
+DATA_PATH = r"E:\pycharm\peptide_data\multitask_peptidepedia\combine.csv"
```
[the remark this answers is not in the document]
Comment (Zoe2021Wang) [reply src/pepbenchmark/pep_utils/neg_sample.py]: 好的，这里现在暂时改成了从仓库下载，之后也避免使用绝对路径

File: src/pepbenchmark/metadata.py
Comment (zjiahui599) [reply src/pepbenchmark/metadata.py:1]: 1. 将1 改为None; 2.暂时不修改，看看后续是否需要区分 两种都有的情况

File: src/pepbenchmark/single_pred/singeltask_dataset.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (zjiahui599) [reply src/pepbenchmark/single_pred/singeltask_dataset.py]: self.features.values()：拿到所有的 value（就是所有的特征值列表），iter(...)：把这些 value 变成一个可迭代对象
next(...)：从这个迭代器中取第一个 特征value，并计算它的长度。
写一个测试验证一下。

File: src/pepbenchmark/metadata.py
Comment (Kqiii) [reply src/pepbenchmark/metadata.py:1]: > 1. 将1 改为None; 2.暂时不修改，看看后续是否需要区分 两种都有的情况

第一个问题，再加一个key呢？
"task_type": "classification" / "regression" 
"num_class": Optional[int] # 在"task_type" == "classification"的时候起效. 如果传入是数字，但是type是regression, 就什么都不做或者抛出提示信息也可以.

突然想到一个问题：会有multilabel regression/ classifcation的情况吗？

File: src/pepbenchmark/pep_utils/neg_sample.py [hunk 10]
```diff
@@ -12,94 +12,167 @@
...
-            excluded_activities = [a.lower() for a in mapping]
-            mask = bio_df[excluded_activities].sum(axis=1) == 0
-            neg_pool_df = bio_df[mask][["sequence"]].copy()
-        elif select == "inclusive":
-            included_activities = [a.lower() for a in mapping]
-            mask = bio_df[included_activities].sum(axis=1) >= 1
-            neg_pool_df = bio_df[mask][["sequence"]].copy()
+    def _download_negative_pool(self) -> None:
+        url = "https://raw.githubusercontent.com/ZGCI-AI4S-Pep/peptide_data/main/multitask_peptidepedia.csv"
+        negative_pool_path = os.path.join("negative_pool", "multitask_peptidepedia.csv")
```
Comment (zjiahui599) [inline src/pepbenchmark/pep_utils/neg_sample.py:105]: 数据统一存储到DATA_DIR下
